### ISSUE TITLE
Deterministic/Reproducible Directory Listing

### ISSUE THREAD

**Issue by indolering** [WebAssembly/wasi-filesystem#12]
The order in which files in a directory are listed is usually determined by when they were added to the filesystem b-tree and is a common source of non-determinism.  A deterministic filesystem API would sort directory listings and iteration based on filename.

Unicode provides a local-invariant default sorting algorithm (DUCET), which can be made [deterministic](http://www.unicode.org/reports/tr10/index.html#Deterministic_Sorting). But it is not suitable for use in WASI for several reasons:
* It's a complex algorithm and ripe for subtle bugs.
* Sort order is unstable across versions of Unicode, even to assigned code-points.  So we would have to pick a single version and turn WASI into a snowflake standard.
* Unlike with casefolding, Unicode collation is *heavily* locale dependent.  Any presentation to the user would require recalculating the sort keys, so it's not worth trying to amortize the cost of DUCET by calculating the keys upfront and storing them with the filename.

The consensus on the [ICU support mailing list](https://sourceforge.net/p/icu/mailman/icu-support/thread/CABWuLVfqWoL%2BfXpuc_GEw%2B4Razh%2BrJk6U272Yq7Y0oVTdhpTTQ%40mail.gmail.com/#msg37136727) was that code-point order is the only sane option: given the high-bar of a portable deterministic sort order, anything more complex is an implementation hazard.  

Plain code-point order would be simplest to implement and the only truly version-less sort order.  However, subtle platform differences may alter the raw bag-of-bytes filename encoding:  `Å` could be encoded as the nordic letter `U+00C5`, the symbol `U+212B`, or the ASCII letter `A`/`U+0041` followed by "COMBINING RING ABOVE" `°`/`030A`.  Non-exhaustive list of things that can alter filename encoding:
* Differences in normalization handling in the filestore (HFS+, SMB, [Git](https://github.com/git/git/commit/3a59e5954ef19ac94522219c2f29d49a187d31d8), etc).
* Differences in input methods, user locale, and available Unicode version.
* Differences between programming languages.
* Complications arising from ensuring round-trip encoding between legacy code-pages and Unicode.

The "solution" is to perform the initial sort on some maximally compatible canonical normalization, using progressively less extreme normal forms (and eventually the raw byte-string) as tie-breakers.  As the file store ultimately decides whether two strings are equivalent in a given namespace, this would result in an identical sort or raising an error. ~~However, the idiot-in-a-hurry implementation of this "solution"  (i.e. NFKx_casefold -> NFKx -> NFx -> WTF-8 byte-string -> UTF-8 byte-string) introduces its own set of stability, complexity, and performance issues.~~

TL;DR
=====
0. Top level code-point sort based on a maximally compatible canonical normalization/lowest common denominator normalization to remove platform dependent differences.
1. Straight NFC tie-breaker.  This is roughly at the level of what most file stores use, except for edge cases like default ignorables and fractions.
2. Raw binary tie-breaker.  Again, the design intentionally minimizes the need for this, as differences due to UCS-2/WTF-8 make these comparisons non-nonsensical/non-portable.

I'm hesitant to suggest any further refinements without implementer feedback.

**Comment by indolering** [comment]
I don't think a case-folded comparison is necessary for deterministic sort order, as only ancient filesystems do not preserve case.  It theoretically reduces entropy in some silly edge cases, but it is probably unhelpful in practice. I would only add a casefolding comparison level as an option if someone from a reproducible build community says it would be helpful or we have to store NFKC_casefolded form anyway.

**Comment by indolering** [comment]
The choice between Composed (`Å`) vs Decomposed (`A` + `°`) canonical [Normal Forms](https://unicode.org/reports/tr15/) (NFC vs NFD) is somewhat arbitrary, as `NFC(NFD(string)) == NFC(string)`.  NFD is often used in filesystems (HFS+ and Linux) but the IETF i18n directorate [thinks this is a mistake](https://mailarchive.ietf.org/arch/search/?q=%22draft-williams-filesystem-18n%22%20nfc).  The vast majority of content on the web and input through keyboards is in NFC form anyway, so it is probably the best choice here as well.

**Comment by indolering** [comment]
I was hoping to skip the NFKx level and define a single ~NFC-superset level comparison that would ignore the differences in normalization use in popular filesystems.  However, this turned out to be fairly brittle and I found out that different input methods may insert strings that won't be identical even in NFx form (i.e. [Korean](https://github.com/rust-lang/rust/issues/2253#issuecomment-29050949): "same keypress will result in U+3131 in Windows but U+1100 in Mac").

**Comment by indolering** [comment]
Lowest Common Denominator
=======================
This requires implementer feedback.

The simplest to implement is `NFKC_Casefold`, which removes default ignorables and is available as an optimized function in the ICU and a character mapping file in the UCD.  However, control codes are **not** default ignore-able and it casefolds, which is unnecessary.  This also removes the non-zero width joiner, which is necessary in some scripts.

We could do NFKC - default ignorables - C1 - C0 (ASCII control codes).  This would be the least destructive but also a snowflake.

We could also use the [PRECIS](https://tools.ietf.org/html/rfc5892#section-2.8) NFKC Freeform profile.  If we use PRECIS in other parts of the standard, then it would make sense to use it here.  This has the benefit of additional controls around the ZWJ and bidirectional text.  It would, however, remove unassigned code points.

**Comment by sunfishcode** [comment]
It would seem that anything that would incorporate NFC, NFD, or Casefold at any part of the ordering would seem to depend on a version of Unicode, and thus be nondeterministic.

Also, as I detailed [here](https://github.com/WebAssembly/wasi-filesystem/issues/5#issuecomment-962343948), I expect WASI-filesystem will contain a significant amount of nondeterminism in practice, so making this one aspect deterministic doesn't seem worth the significant complexity here.

**Comment by indolering** [comment]
> It would seem that anything that would incorporate NFC, NFD, or Casefold at any part of the ordering would seem to depend on a version of Unicode, and thus be nondeterministic.

Reproducibility is the best any platform can hope to accomplish, as any Turing complete program with access to non-deterministic inputs can be made to run non-deterministically.  Sorting by codepoint would not be version dependent, cover ~95% of use cases, and would be a LOT less work than implementing a VFS or ad-hoc patching to sort filesystem lookups based on file name.

Point taken WRT complexity, performance, and implementation costs.  I was hoping to propose an Evergreen Unicode release strategy and contribute a sorting implementation, but medical issues have delayed this work.  Ideally, things would be architected to allow someone with the budget/time to contribute a maximally normalizing sort algorithm later on.

But feel free to close and tag with `wontfix` if that is deemed out-of-scope.  We've already spent a lot of time in this rabbit hole.

**Comment by indolering** [comment]
Yeah, let's kill this.  Without the FS maintaining the index, we will get weird performance edge cases.  It's just the wrong level of abstraction to be doing this on.
